### PR TITLE
Python: container summaries, part 3

### PR DESCRIPTION
**Pull request by yoff**
In which we remove the final explicit taint steps from `containerStep`.

### CODE REVIEW

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 1]
```diff
@@ -4127,6 +4235,173 @@ private module StdlibPrivate {
...
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      // existing elements
+      input = "Argument[self].ListElement" and
+      output = "ReturnValue.ListElement" and
+      preservesValue = true
+      or
+      // newly added element returned
+      input = "Argument[0]" and
+      output = "ReturnValue.ListElement" and
+      preservesValue = true
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4345]: I don't see `list.append` having a return value. What are these flows about? :confused:

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 2]
```diff
@@ -4127,6 +4235,173 @@ private module StdlibPrivate {
...
+      or
+      // transfer taint from new element to this
+      input = "Argument[0]" and
+      output = "Argument[self]" and
+      preservesValue = false
+      or
+      // transfer taint from new element to return value
+      input = "Argument[0]" and
+      output = "ReturnValue" and
+      preservesValue = false
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4360]: Also this

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 3]
```diff
@@ -4127,6 +4235,173 @@ private module StdlibPrivate {
...
+      or
+      // newly added element added to this
+      input = "Argument[0]" and
+      output = "Argument[self].ListElement" and
+      preservesValue = true
+      or
+      // transfer taint from new element to this
+      input = "Argument[0]" and
+      output = "Argument[self]" and
+      preservesValue = false
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4355]: I assume we don't want this forever?

```suggestion
      // transfer taint from new element to this (TODO: remove in future when taint-handling is more in line with other languages)
      input = "Argument[0]" and
      output = "Argument[self]" and
      preservesValue = false
```

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 4]
```diff
@@ -4127,6 +4235,173 @@ private module StdlibPrivate {
...
+      // existing elements
+      input = "Argument[self].SetElement" and
+      output = "ReturnValue.SetElement" and
+      preservesValue = true
+      or
+      // newly added element returned
+      input = "Argument[0]" and
+      output = "ReturnValue.SetElement" and
+      preservesValue = true
+      or
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4388]: Same as above, set.add doesn't seem to have a return value when I try it out in a real `ipython` scenario

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 5]
```diff
@@ -4127,6 +4235,173 @@ private module StdlibPrivate {
...
+      // transfer taint from new element to this
+      input = "Argument[0]" and
+      output = "Argument[self]" and
+      preservesValue = false
+      or
+      // transfer taint from new element to return value
+      input = "Argument[0]" and
+      output = "ReturnValue" and
+      preservesValue = false
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4402]: same

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 6]
```diff
@@ -4127,6 +4235,173 @@ private module StdlibPrivate {
...
+      or
+      // newly added element added to this
+      input = "Argument[0]" and
+      output = "Argument[self].SetElement" and
+      preservesValue = true
+      or
+      // transfer taint from new element to this
+      input = "Argument[0]" and
+      output = "Argument[self]" and
+      preservesValue = false
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4397]:
```suggestion
      // transfer taint from new element to this (TODO: remove in future when taint-handling is more in line with other languages)
      input = "Argument[0]" and
      output = "Argument[self]" and
      preservesValue = false
```

File: python/ql/lib/change-notes/2023-06-13-container-store-steps.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* More precise modelling of several container functions and methods.
```
Comment (RasmusWL) [inline python/ql/lib/change-notes/2023-06-13-container-store-steps.md:4]: This change note is a bit abstract, so I've suggested some concrete examples

```suggestion
* More precise modelling of several container functions (such as `sorted`, `reversed`) and methods (such as `set.add`, `list.append`).
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Really nice to see so many `MISSING:` being fixed :muscle: :tada: (just a few questions that needs to be resolved)

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll (hunk 1 above)
Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Stdlib.qll:4345]: Hm, not sure where I got this from. Neither the linked documentation nor a local experiment agrees with this..

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 8]
```diff
@@ -4127,6 +4235,153 @@ private module StdlibPrivate {
...
+    override DataFlow::ArgumentNode getACallback() {
+      result.(DataFlow::AttrRead).getAttributeName() = "append"
+    }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      // existing elements
+      input = "Argument[self].ListElement" and
+      output = "ReturnValue.ListElement" and
+      preservesValue = true
+      or
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4341]: Seems like you missed these two?

```suggestion
```

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll [hunk 9]
```diff
@@ -4127,6 +4235,153 @@ private module StdlibPrivate {
...
+    override DataFlow::ArgumentNode getACallback() {
+      result.(DataFlow::AttrRead).getAttributeName() = "add"
+    }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      // existing elements
+      input = "Argument[self].SetElement" and
+      output = "ReturnValue.SetElement" and
+      preservesValue = true
+      or
```
Comment (RasmusWL) [inline python/ql/lib/semmle/python/frameworks/Stdlib.qll:4373]: Seems like you missed these two?

```suggestion
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]

File: python/ql/lib/semmle/python/frameworks/Stdlib.qll (hunk 8 above)
Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Stdlib.qll]: Indeed, thanks :-)

**Review by RasmusWL — APPROVED** [review]

**Pull request merged**